kura/examples/test/org.eclipse.kura.example.wire.trigonometric.functions.provider.test/src/main/java/org/eclipse/kura/example/wire/trigonometric/functions/provider/test/TrigonometricComponentTest.java

```diff
@@ -0,0 +1,348 @@
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ *******************************************************************************/
+
+package org.eclipse.kura.example.wire.trigonometric.functions.provider.test;
+
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.kura.KuraException;
+import org.eclipse.kura.configuration.ComponentConfiguration;
+import org.eclipse.kura.configuration.ConfigurationService;
+import org.eclipse.kura.type.TypedValue;
+import org.eclipse.kura.type.TypedValues;
+import org.eclipse.kura.util.wire.test.GraphBuilder;
+import org.eclipse.kura.util.wire.test.TestEmitterReceiver;
+import org.eclipse.kura.util.wire.test.WireTestUtil;
+import org.eclipse.kura.wire.WireEnvelope;
+import org.eclipse.kura.wire.WireRecord;
+import org.eclipse.kura.wire.graph.WireGraphConfiguration;
+import org.eclipse.kura.wire.graph.WireGraphService;
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.service.component.ComponentContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TrigonometricComponentTest {
+
+    // See:
+    // http://stackoverflow.com/questions/7161338/using-osgi-declarative-services-in-the-context-of-a-junit-test
+
+    private static final Logger logger = LoggerFactory.getLogger(TrigonometricComponentTest.class);
+
+    private static final String UNDER_TEST_PID = "under.test";
+    private static final String TEST_EMITTER_PID = "test.emitter.pid";
+    private static final String TEST_RECEIVER_PID = "test.receiver.pid";
+
+    // in ports
+    private static final int IN_PORT = 0;
+
+    // out ports
+    private static final int OUT_PORT = 0;
+
+    // configuration properties of component under test
+    private static final String OPERAND_NAME_PROP_NAME = "operand.name";
+    private static final String RESULT_NAME_PROP_NAME = "result.name";
+    private static final String EMIT_RECEIVED_PROPERTIES = "emit.received.properties";
+    private static final String TRIGONOMETRIC_OPERATION = "trigonometric.operator";
+
+    private static final String OPERAND_NAME_DEFAULT = "operand";
+    private static final String RESULT_NAME_DEFAULT = "result";
+    private static final boolean EMIT_RECEIVED_PROPERTIES_DEFAULT = false;
+    private static final String TRIGONOMETRIC_OPERATION_DEFAULT = "SIN";
+
+    // tolerance for errors with inverse trigonometric functions
+    private static final double ACCEPTABLE_ERROR = 0.0000000000000004;
```

Comment:
Can you please explain this?

Reply:
Because of how floating point values are treated, and because of approximations, results of Math functions can slightly differ from the real values.
Different functions lead to slightly different errors, all less than the ACCEPTABLE_ERROR value.

If you take SIN of PI, expected result should be 0, while result is off slightly (1.2246467991473532E-16).

For tests' sake, instead of rounding, or performing the same mathematical operation with java.lang.Math and checking the value (IMHO pretty much useless), I preferred matching the values to the expected values of plus/minus one, zero or Pi fractions (using Math.PI), taking in account of this error.

Reply:
The accuracy of the java.lang.Math methods is documented (see https://docs.oracle.com/javase/8/docs/api/java/lang/Math.html#sin-double-)

> or performing the same mathematical operation with java.lang.Math and checking the value (IMHO pretty much useless)

why useless?

Reply:
First, to use the output values (like the 1.2246467991473532E-16 "zero" from my previous comment) is really not human friendly. 
If we consider that tests most of the time have the side effect of showing how the associated bundle works, human readibility in my opinion is important.

Second, by my taste to have something like the following is quite ugly:

> Double result = <output of the component, e.g. SIN of Pi)>
assertTrue(Math.sin(Math.PI) == result);

Let me know how you want it done.